addons/account/models/account_journal_dashboard.py

```diff
@@ -257,178 +261,210 @@ def _get_bar_graph_select_query(self):
-            'nb_lines_bank_account_balance': nb_lines_bank_account_balance,
-            'outstanding_pay_account_balance': formatLang(self.env, currency.round(outstanding_pay_account_balance), currency_obj=currency),
-            'nb_lines_outstanding_pay_account_balance': nb_lines_outstanding_pay_account_balance,
-            'last_balance': formatLang(self.env, currency.round(last_balance) + 0.0, currency_obj=currency),
-            'number_draft': number_draft,
-            'number_waiting': number_waiting,
-            'number_late': number_late,
-            'sum_draft': formatLang(self.env, currency.round(sum_draft) + 0.0, currency_obj=currency),
-            'sum_waiting': formatLang(self.env, currency.round(sum_waiting) + 0.0, currency_obj=currency),
-            'sum_late': formatLang(self.env, currency.round(sum_late) + 0.0, currency_obj=currency),
-            'currency_id': currency.id,
-            'bank_statements_source': self.bank_statements_source,
-            'title': title,
-            'is_sample_data': is_sample_data,
-            'company_count': len(self.env.companies)
+    def _fill_sale_purchase_dashboard_data(self, dashboard_data):
+        sale_purchase_journals = self.filtered(lambda journal: journal.type in ('sale', 'purchase'))
+        if not sale_purchase_journals:
+            return
+        field_list = [
+            "account_move.journal_id",
+            "(CASE WHEN account_move.move_type IN ('out_refund', 'in_refund') THEN -1 ELSE 1 END) * account_move.amount_residual AS amount_total",
+            "account_move.amount_residual_signed AS amount_total_company",
+            "account_move.currency_id AS currency",
+            "account_move.move_type",
+            "account_move.invoice_date",
+            "account_move.company_id",
+        ]
+        self.env.cr.execute(*sale_purchase_journals._get_open_bills_to_pay_query().select(*field_list))
+        query_results_to_pay = group_by_journal(self.env.cr.dictfetchall())
+
+        self.env.cr.execute(*sale_purchase_journals._get_draft_bills_query().select(*field_list))
+        query_results_drafts = group_by_journal(self.env.cr.dictfetchall())
+
+        self.env.cr.execute(*sale_purchase_journals._get_late_bills_query().select(*field_list))
+        late_query_results = group_by_journal(self.env.cr.dictfetchall())
+
+        to_check_vals = {
+            vals['journal_id']: vals
+            for vals in self.env['account.move'].read_group(
+                domain=[('journal_id', 'in', sale_purchase_journals.ids), ('to_check', '=', True)],
+                fields=['amount_total'],
+                groupby='journal_id',
+                lazy=False,
+            )
+        }
+
+        curr_cache = {}
+        sale_purchase_journals._fill_dashboard_data_count(dashboard_data, 'account.move', 'entries_count', [])
+        for journal in sale_purchase_journals:
+            currency = journal.currency_id or journal.company_id.currency_id
```

Comment:
I believe that journal.currency_id doesn't have any sense here, although it appears (I check in Odoo 16) you can effectively set a secondary currency on a sale/purchase journal. :upside_down_face: 
I don't think it has any interest, except maybe to default the secondary currency of invoices created in that journal, and reporting in this very dashboard... I saw it was like that before, but we should probably consider simplifying that, what do you think? 


edit: you use amount_residual which is in company currency to compute amount_total, so... yeah :-) We should clean that